core/drivers/imx_ele.c

```diff
@@ -0,0 +1,576 @@
+} __packed;
+
+struct session_get_device_info_rsp {
+	uint32_t rsp_code;
+	uint32_t user_sab_id;
+	uint32_t chip_uid[4];
+	uint16_t chip_life_cycle;
+	uint16_t chip_monotonic_counter;
+	uint32_t ele_version;
+	uint32_t ele_version_ext;
+	uint8_t fips_mode;
+	uint8_t reserved[3];
+	uint32_t crc;
+} __packed;
+
+struct response_code {
+	uint8_t status;
+	uint8_t rating;
+	uint16_t rating_extension;
+} __packed;
+
+/*
+ * Print ELE response status and rating
+ *
+ * @rsp response code structure
+ */
+static void print_rsp_code(struct response_code rsp __maybe_unused)
+{
+	DMSG("Response status %#"PRIx8", rating %#"PRIx8" (ext %#"PRIx16")",
+	     rsp.status, rsp.rating, rsp.rating_extension);
+}
+
+/*
+ * Print ELE message header
+ *
+ * @hdr message header
+ */
+static void print_msg_header(struct imx_mu_msg_header hdr __maybe_unused)
+{
+	DMSG("Header vers %#"PRIx8", size %"PRId8", tag %#"PRIx8
+	     ", cmd %#"PRIx8,
+	     hdr.version, hdr.size, hdr.tag, hdr.command);
+}
+
+/*
+ * Print full ELE message content
+ *
+ * @msg message
+ */
+static void dump_message(const struct imx_mu_msg *msg __maybe_unused)
+{
+	size_t i = 0;
+	size_t size = msg->header.size;
+	uint32_t *data __maybe_unused = (uint32_t *)msg;
+
+	DMSG("Dump of message %p(%zu)", data, size);
+	for (i = 0; i < size; i++)
+		DMSG("word %zu: %#"PRIx32, i, data[i]);
+}
+
+/*
+ * Return the number of 32 bits words of the given message.
+ *
+ * @cmd command size in byte
+ */
+static size_t size_msg_32(size_t cmd)
+{
+	/* Roundup and add header size */
+	return ROUNDUP_DIV(cmd, sizeof(uint32_t)) + 1;
+}
+
+/*
+ * The CRC for the message is computed xor-ing all the words of the message:
+ * the header and all the words except the word storing the CRC.
+ *
+ * @msg MU message to hash
+ */
+static uint32_t compute_crc(const struct imx_mu_msg *msg)
+{
+	uint32_t crc = 0;
+	uint8_t i = 0;
+	uint32_t *payload = (uint32_t *)msg;
+
+	assert(msg);
+
+	for (i = 0; i < msg->header.size - 1; i++)
+		crc ^= payload[i];
+
+	return crc;
+}
+
+/*
+ * The CRC is the last word of the message
+ *
+ * @msg MU message to hash
+ */
+static void update_crc(struct imx_mu_msg *msg)
+{
+	assert(msg);
+	/*
+	 * The CRC field is the last element of array. The size of the header
+	 * is also subtracted from CRC computation.
+	 */
+	msg->data.u32[msg->header.size - 2] = compute_crc(msg);
+}
+
+/*
+ * Return the given MU base address, depending on the MMU state.
+ *
+ * @pa MU physical base address
+ * @sz MU size
+ */
+static vaddr_t imx_ele_init(paddr_t pa, size_t sz)
+{
+	static bool is_initialized;
+	vaddr_t va = 0;
+
+	assert(pa && sz);
+
+	if (cpu_mmu_enabled())
+		va = core_mmu_get_va(pa, MEM_AREA_IO_SEC, sz);
+	else
+		va = (vaddr_t)pa;
+
+	if (!is_initialized) {
+		imx_mu_init(va);
+		is_initialized = true;
+	}
+
+	return va;
+}
+
+/*
+ * Extract response codes from the given word
+ *
+ * @word 32 bits word MU response
+ */
+static struct response_code get_response_code(uint32_t word)
```

Comment:
Is it ok to return a struct? Does compiler checks struct matches the architecture return value native size?

Reply:
Depending of the struct size, the compiler might handle it differently. But yes, returning a struct makes sense.